### DISCUSSION TITLE
[C++ Networking] Http Nginx Keepalives

### DISCUSSION THREAD

**Discussion by DelinWorks** [axmolengine/axmol#612] [Q&A]
I've set up an nginx server so I can host my login and signup system for this game and load the scores

Problem is that I've told nginx to keep the connections open for 120 seconds and it seems that whenever I send a request through my game the connection gets released and not kept alive. I've known this because chrome keeps the connection for exactly 120 secs using NetLimiter and when I inspect the game using NetLimiter the game just releases the connection as soon as the request is done

Why would I keep connections alive? it's because when you're working with multiplayer games you want to minimize the amount of tcp connections to be initialized. so that the server and the game feel fast

As you can see chrome keeps the connection alive as said in nginx config:

![image](https://user-images.githubusercontent.com/45469625/160904305-ace01e1b-f8b9-449f-8eae-41f74a9966ac.png)

But the game keeps dropping the connection as soon as the request is finished:

![image](https://user-images.githubusercontent.com/45469625/160904424-1a5d37b2-46c5-47dd-9ab1-003f6547da1b.png)

So how would you go about implementing this in c++ ? (Keeping connections alive for later use)

I initiate connections this way

```
auto request = new (std::nothrow) network::HttpRequest();
request->setRequestType(network::HttpRequest::Type::GET);
request->setUrl("https://delingames.xyz/accounts/load_score.php?uid=1");
request->setResponseCallback([&](network::HttpClient* client, network::HttpResponse* response) {
    if (response->isSucceed()) {
        // process request from server
    });
network::HttpClient::getInstance()->send(request);
```

**Comment by aismann** [comment]
@DelinWorks 
Any news? Should be deleted? Or add as a feature to axmol?

**Comment by rh101** [comment]
@halx99 As has been asked here by delin, is there any way to keep the HTTP connection alive?

I'm seeing this in the console output for all HTTP requests:
```
[yasio][1706800445027][index: 0] connecting server somedomain.net(*.*.*.*):***...
[yasio][1706800445037][index: 0] the connection #28 <X.X.X.X:*> --> <*.*.*.*:***> is established.
[yasio][1706800445150][index: 0] the connection #28 is lost, ec=-24, where=0, detail:An existing connection was shutdown by local host!
```

Does the current implementation using yasio instantly shut down the connection once the HTTP response is received? Is there any way to configure it (per request) so it does keep the connection alive? At the moment it seems to be going through the entire handshake sequence for both TCP and SSL per request, which I assume can be avoided if we could re-use the same connection.

**Reply by halx99** [reply]
It's possible keepalive, but the logic may more complex

**Reply by DelinWorks** [reply]
this would really help in environments where renting a vps server has limited bandwidth per month, no need for extra bandwidth and wasted packets initiating and closing tcp connections (SYN, ACK, FIN, ACK), especially if the data is frequent informational json data and not blobs

**Reply by rh101** [reply]
I wasn't noticing the issue in my local testing setup, and there was only a small (approx. 10ms) difference between HTTP and HTTPS connection response times (due to the SSL handshake). 

My issue now is that the further away the server is, the worse the problem will be, as the TCP and SSL connection handshakes take so much longer to complete, and it would add significant latency to repeated requests to the same endpoint.  I really wish I noticed this earlier.

**Reply by rh101** [reply]
@halx99 Should we create an issue to track this?  That is assuming it's something you're willing to implement if you have the time.

**Reply by halx99** [reply]
The issue #1651 has been created

**Comment by 04pmsuu1** [comment]
I havent tried to reprocude, but did you try to setup network service with YOPT_S_TCP_KEEPALIVE?

**Reply by halx99** [reply]
The `YOPT_S_TCP_KEEPALIVE` not keep connection, it's enable tcp internal heartbeat detection mechanism
